cpp/src/arrow/compute/kernels/vector_rank.cc

```diff
@@ -381,9 +381,13 @@ class RankMetaFunction : public RankMetaFunctionBase<RankMetaFunction> {
   }
 
   RankMetaFunction()
-      : RankMetaFunctionBase("rank", Arity::Unary(), rank_doc, &kDefaultOptions) {}
+      : RankMetaFunctionBase("rank", Arity::Unary(), rank_doc, GetDefaultOptions()) {}
```

Comment:
@pitrou well it's not *one* line, but...

```suggestion
      : RankMetaFunctionBase("rank", Arity::Unary(), rank_doc, [] {
          static const auto kDefaultOptions = RankOptions::Defaults();
          return &kDefaultOptions;
        }()) {}
```

An IIFE can introduce a static lifetime object too

Reply:
IIFE?

Reply:
Sorry, [Immediately Invoked Function Expression](https://en.wikipedia.org/wiki/Immediately_invoked_function_expression). I don't believe it's an official C++ concept

Reply:
Do you think it would work to do something like:
```c++
// in .h file
template <typename OptionsType>
const FunctionOptions* StaticOptionsInit() {
  static const auto kDefaultOptions = OptionsType::Defaults();
  return &kDefaultOptions
};

// in .cc file
class RankMetaFunction : public RankMetaFunctionBase<RankMetaFunction> {
 public:
  RankMetaFunction()
      : RankMetaFunctionBase("rank", Arity::Unary(), rank_doc, StaticOptionsInit<RankOptions>()) {}
```

Reply:
An inline variable will have the same issue that this PR tries to fix, right? i.e. it could be initialized before the corresponding function options type.

Reply:
> Also `StaticOptionsInit` is neat, and maybe we should extend it to be

Given the LLVM issue you linked to above, I'm now convinced we shouldn't do something like this :)

Reply:
Finally, what approach should we use here? :-)

Reply:
Sorry @kou. I think we can use either `GetDefaultOptions` or the one suggested in https://github.com/apache/arrow/pull/46303#discussion_r2073899428 . At your preference :)

Reply:
OK. Let's choose `GetDefaultOptions()` here. `GetDefaultOptions()` is more straightforward than https://github.com/apache/arrow/pull/46303#discussion_r2073899428 because developers don't need to know IIFE and static variable lifetime.